hypervisor/arch/x86/notify.c

```diff
@@ -8,15 +8,58 @@
 
 static struct dev_handler_node *notification_node;
 
+static volatile uint64_t smp_call_mask = 0UL;
+
 /* run in interrupt context */
 static int kick_notification(__unused int irq, __unused void *data)
 {
-	/* Notification vector does not require handling here, it's just used
-	 * to kick taget cpu out of non-root mode.
+	/* Notification vector is used to kick taget cpu out of non-root mode.
+	 * And it also serves for smp call.
 	 */
+	uint16_t pcpu_id = get_cpu_id();
+
+	if (bitmap_test(pcpu_id, &smp_call_mask)) {
+		struct smp_call_info_data *smp_call =
+			&per_cpu(smp_call_info, pcpu_id);
+
+		if (smp_call->func)
+			smp_call->func(smp_call->data);
+		bitmap_clear_nolock(pcpu_id, &smp_call_mask);
+	}
+
 	return 0;
 }
 
+void smp_call_function(uint64_t mask, smp_call_func_t func, void *data)
+{
+	uint16_t pcpu_id;
+	struct smp_call_info_data *smp_call;
+
+	/* wait for previous smp call complete, which may run on other cpus */
+	wait_sync_change(&smp_call_mask, 0UL);
```

Comment:
This is redundant?

Reply:
no, it's used to sync smp_call which may come from other cpus when one cpu is running on it.

Reply:
if  you use smp_call_mask==0 to indicate the caller can start a new SMP call, that is dangerous

Reply:
Ok, so you mean multi cpus could do it at exact same time, right? Yes, as wait_sync_change & atomic_store64 is not an atomic operation, I am doing it just want to avoid to use a lock, if so, I will add the lock back.

Reply:
Not necessary to use lock, but make sure you use test_and_exchange atmoic operation.

Reply:
Sure.

Reply:
atomic_cmpxchg64